### PR TITLE
kbfs sagas

### PR DESCRIPTION
**Pull request by chrisnojima**
This sagaizes the kbfs actions

### CODE REVIEW

File: shared/actions/kbfs/index.js [hunk 1]
```diff
@@ -1,33 +1,47 @@
 // @flow
 import * as Constants from '../../constants/kbfs'
-import type {AsyncAction} from '../../constants/types/flux'
-import {fsListRpc} from '../../constants/types/flow-types'
-import {openInKBFS as platformOpenInKBFS} from './index.platform'
+import {fsListRpcPromise} from '../../constants/types/flow-types'
+// $FlowIssue flow doens't understand generator exports
+import {openSaga} from './index.platform'
```
Comment (chrisnojima) [inline shared/actions/kbfs/index.js]: unclear why flow doesn't understand this

File: shared/actions/kbfs/index.js [hunk 2]
```diff
@@ -1,33 +1,47 @@
...
+
+function fsList (path: string) : FSList {
+  return {type: Constants.fsList, payload: {path}}
+}
+
+function openInKBFS (path: string = ''): FSOpen {
+  return {type: Constants.fsOpen, payload: {path}}
+}
+
+function * _listSaga (action: FSList): SagaGenerator<any, any> {
```
Comment (chrisnojima) [inline shared/actions/kbfs/index.js]: this isn't actually consumed by anything (as before). I think @gabriel added this during the kbfs mobile stuff so its not really hooked up / used anywhere yet

File: shared/actions/kbfs/index.platform.desktop.js [hunk 3]
```diff
@@ -48,55 +49,70 @@ function formKbfsPathWindows (extendedConfig: ExtendedStatus): string {
...
+        kbfsPath = yield call(windowsKBFSRoot, extendedConfig)
+        if (kbfsPath) {
+          break
+        }
+      } catch (error) {
+        kbfsPathError = error
+      }
+
+      yield put(getExtendedStatus())
+      yield race({
```
Comment (chrisnojima) [inline shared/actions/kbfs/index.platform.desktop.js]: race so we don't get stuck here forever

File: shared/actions/kbfs/index.platform.desktop.js [hunk 4]
```diff
@@ -48,55 +49,70 @@ function formKbfsPathWindows (extendedConfig: ExtendedStatus): string {
...
-      open(path.resolve(kbfsPath, openPath))
+
+    yield put({type: Constants.changeKBFSPath, payload: {path: kbfsPath}})
+  }
+
+  if (kbfsPath) {
+    openPath = path.resolve(kbfsPath, openPath)
+    // Check to make sure our resolved path starts with the kbfsPath
+    // i.e. (not opening a folder outside kbfs)
+    if (!openPath.startsWith(kbfsPath)) {
```
Comment (chrisnojima) [inline shared/actions/kbfs/index.platform.desktop.js]: this check happens on both logical paths now instead of just one as before

File: shared/actions/kbfs/index.platform.desktop.js [hunk 5]
```diff
@@ -48,55 +49,70 @@ function formKbfsPathWindows (extendedConfig: ExtendedStatus): string {
...
 }
 
-export function openInKBFS (openPath: string = Constants.defaultKBFSPath): AsyncAction {
+function * openSaga (action: FSOpen): SagaGenerator<any, any> {
+  const openPath = action.payload.path || Constants.defaultKBFSPath
+
   console.log('openInKBFS:', openPath)
   if (process.platform === 'win32') {
-    return openInWindows(openPath)
+    yield * openInWindows(openPath)
```
Comment (chrisnojima) [inline shared/actions/kbfs/index.platform.desktop.js]: doesn't seem to work correctly with yield * call(openInWindows, openPath), unclear if thats correct or not. this does work... @MarcoPolo ?

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready to look at

File: shared/actions/kbfs/index.platform.desktop.js (hunk 5 above)
Comment (MarcoPolo) [inline shared/actions/kbfs/index.platform.desktop.js]: It should just be a `yield call(openInWindows, openPath)` but this works too.

`yield * call(...)` doesn't make sense because call doesn't return a generator.

File: shared/search/index.js [hunk 6]
```diff
@@ -31,7 +32,7 @@ class Search extends Component<void, Props, void> {
   }
 }
 
-const connector: TypedConnector<TypedState, TypedDispatch<SearchActions>, OwnProps, Props> = new TypedConnector()
+const connector: TypedConnector<TypedState, TypedDispatch<SearchActions | FSOpen>, OwnProps, Props> = new TypedConnector()
```
Comment (MarcoPolo) [inline shared/search/index.js]: Was this needed?

File: shared/actions/kbfs/index.platform.js.flow [hunk 7]
```diff
@@ -1,6 +1,4 @@
 // @flow
-import type {AsyncAction} from '../../constants/types/flux'
+import type {Action} from '../../constants/types/flux'
 
-export function openInKBFS (path: string = ''): AsyncAction {
-  return (dispatch, getState) => {}
-}
+declare function openInKBFS (path: ?string): Action
```
Comment (MarcoPolo) [inline shared/actions/kbfs/index.platform.js.flow]: This should be type `FSOpen`

File: shared/actions/kbfs/index.js (hunk 1 above)
Comment (chrisnojima) [inline shared/actions/kbfs/index.js]: doh!

File: shared/actions/kbfs/index.platform.desktop.js (hunk 5 above)
Comment (chrisnojima) [inline shared/actions/kbfs/index.platform.desktop.js]: i thought if the inner thing generates generators i needed to defer with the *??

File: shared/search/index.js (hunk 6 above)
Comment (chrisnojima) [inline shared/search/index.js]: yes, cause it dispatches this and FSOpen isn't part of the searchactions

Comment (chrisnojima) [inline shared/search/index.js]: i would explore the openInWindows changes if i didn't have to load windows/virtualbox... :)

**Comment by chrisnojima** [comment]
@MarcoPolo updated

File: shared/actions/kbfs/index.platform.desktop.js (hunk 5 above)
Comment (MarcoPolo) [inline shared/actions/kbfs/index.platform.desktop.js]: http://yelouafi.github.io/redux-saga/docs/api/index.html#callfn-args

When you do `call` you aren't actually running the function right away, you are telling the middleware to do it, and it behaves in the way documented above. So it lets the called generator finish before resuming.

I found these docs that explain this better:
http://yelouafi.github.io/redux-saga/docs/advanced/SequencingSagas.html
http://yelouafi.github.io/redux-saga/docs/advanced/ComposingSagas.html

tl;dr yield* works but it's more composable to do `call`
